### PR TITLE
Add clone submodules parameter to CloneGitRepositoryJob

### PR DESCRIPTION
**Pull request by mmz33**
Add a flag to clone submodules if needed. In my case, I had to use repos from `returnn/extern`.

### CODE REVIEW

**Review by michelwi — COMMENTED** [review]
I am wondering if it would be better to have the submodule init call after the specific commit checkout call.

What would happen if some submodules were updated between `commit` and `HEAD` in either case?

**Comment by mmz33** [comment]
@michelwi Ah I thought that the commits are independent. So what do you suggest?

**Comment by michelwi** [comment]
> @michelwi Ah I thought that the commits are independent. So what do you suggest?

I genuinely don't know how/if this works. I imagined that e.g. at commit `123` the repo is
```
repo/
  submodule1
  submodule2@version1
  own_module
```
and then it is later updated to `HEAD`
```
repo/
  submodule2@version2
  own_module
```

Now in the current implementation if you run `CloneGitRepositoryJob("repo", commit="123", clone_submodules=True)`
first the `HEAD` is checked out, then the submodules are initialized (i.e. `submodule2@version2`) and then the specific commit `123` is checked out.
My question is: will `submodule1` now also be prepared? And will `submodule2`'s version be moved to `version1`?

**Comment by mmz33** [comment]
ok so this can be fixed by doing the submodule cloning at the end after checking out the commit. this should work i think.

File: tools/git.py [hunk 1]
```diff
@@ -38,6 +41,11 @@ def run(self):
         args += [repository_dir]
         logging.info("running command: %s" % " ".join(args))
         sp.run(args, check=True)
+
+        if self.clone_submodules:
```
Comment (christophmluscher) [inline tools/git.py:45]: IMHO it would be easier to directly call:

`git clone --recurse-submodules`

Comment (christophmluscher) [reply tools/git.py:45]: although this might run into the same issue as mentioned by Willi

Comment (michelwi) [reply tools/git.py:45]: I did some research and the same flag also exists for checkout

> ```git checkout --recurse-submodules```
Using --recurse-submodules will update the content of all initialized submodules according to the commit recorded in the superproject.
If local modifications in a submodule would be overwritten the checkout will fail unless -f is used.

so you could change the PR in a way that if the flag `clone_submodules` is set, then `--recurse-submodules` is added both to clone and checkout.

**Review by JackTemaki — APPROVED** [review]
I verified with @mmz33 locally that this works.

**Comment by JackTemaki** [comment]
@Atticus1806 thanks a lot

**Review by Atticus1806 — APPROVED** [review]

**Pull request merged**